mesa: move llvmPackages to depsHostHost

oxalica opened the pull request:
###### Motivation for this change

Mesa use LLVM for OpenCL, which is both running on and targeting hostPlatform, thus depsHostHost.

Consider nix-shell with crossSystem for cross platform development, and have some GUI nativeBuildInputs using mesa. These programs are expected to run on buildPlatform, and shouldn't depend on crossSystem, which is buildPackages.mesa's targetPlatform.

I discovered the issue when I put `qemu` in `nativeBuildInputs` of `shell.nix` for cross platform development. And it tries to rebuild `qemu` (build = host = x86_64, target = aarch64) with some aarch64 packages, which is incorrect. Since QEMU's host is x86_64 and it is running on x86_64 to simulate aarch64. No aarch64 artifacts is needed for QEMU and/or its dependency `mesa`.

This PR causes no rebuild for native or cross build. It only makes `pkgsCross.*.buildPackages.mesa == mesa` and `pkgsCross.*.some-pkg == some-pkg` where `some-pkg` relies on `mesa` but doesn't depend on targetPlatform, ~like `qemu`~. Unfortunately recent changes of `qemu` makes it still not target-agnostic, but that are future issues. 

cc: @Mindavi @sbruder  @primeos @vcunat

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

r-burns commented:
Hmm, so does this also mean that these llvm packages should be in mesa's depsHostHost rather than its buildInputs?

oxalica commented:
> so does this also mean that these llvm packages should be in mesa's depsHostHost rather than its buildInputs?

I think so. But it causes evaluation error when building `pkgsCross.aarch64-multiplatform.buildPackages.mesa`, even though I manually specify the package like `depsHostHost = [ pkgsHostHost.* ]`.
```
error: `valueHostHost` unimplemented: pass manually rather than relying on splice.
```
The error is from
https://github.com/NixOS/nixpkgs/blob/e8659f7bf6b850e00c23d88901a028d0b482c263/pkgs/top-level/splice.nix#L45

It seems that `depsHostHost` is not implemented but I'm not sure why. The ancient code was introduced in 2017. :thinking:

r-burns commented:
Huh, that's really weird. The splicing seems to work fine when changing it to match the surrounding lines (i.e. `valueHostHost = pkgsHostHost.${name} or {};`); @Ericson2314 do you remember why that wasn't done initially?

This is indeed the first actual use of depsHostHost in nixpkgs (!) - according to ce2637c97a1d6c59d29fad62678e5f3c261df59a this type of dependency is very rare, cool :)

SuperSandro2000 commented on pkgs/development/libraries/mesa/default.nix:
```diff
@@ -1,4 +1,4 @@
-{ stdenv, lib, fetchurl, fetchpatch, buildPackages
+{ stdenv, lib, fetchurl, fetchpatch, buildPackages, pkgs
```
Do we really need to use pkgs as an input?

oxalica replied:
Do you mean introduce `pkgsCross` and `mesa` here to implement `tests.doesNotRelyOnTargetPlatform` below? That's would be more wired that introducing itself from arguments and also use `self` below.

Note that we cannot write `pkgsCross.aarch64-multiplatform.buildPackages.mesa == self` in the test. Otherwise overriding will definitely break the test. Since we doesn't have something like `makeDrvCrossAarch64 self == self`, my best approach is just the current diff.

Ericson2314 replied:
`pkgsCross` should basically never be used inside packages.

I wish it were in fact impossible (`pkgsCross` either not existing or slapped on after the fact) but there are a few odds and and and old grandfathered in things that I don't have the energy to fight.

r-burns replied:
Hmm, then what's an alternative way to enforce target-invariance here?

oxalica replied:
Is the current test acceptable? Or should I move the test into an individual one of nixosTests?

vcunat replied:
Is there a problem with `pkgs` in the inputs?  I mean, we commonly pass `buildPackages` and similar, so is this really any different?

Ericson2314 commented:
@r-burns I think that can be fixed! :) We have `pkgsHostHost` now so I think it can be implemented normally. Before we didn't, and I think I was spooked by the fact a conditional was involved with the merely linear bootstrapping. And if you look at `pkgs/stdenv/booter.nix`, one can see `pkgsBuildHost` also has the conditionals, so it's not even unique in that regard!

Ericson2314 commented:
What exactly is used from `llvmPackages`? `llvmPackages.llvm` should *also* be target agnostic, so we should start there.

oxalica commented:
> What exactly is used from `llvmPackages`? `llvmPackages.llvm` should _also_ be target agnostic, so we should start there.

`llvmPackages.llvm` is already target agnostic now and doesn't matter here. Currently it's `clang` that matters. And if `depsHostHost` works, we should move all of these pkgs below into it.

https://github.com/NixOS/nixpkgs/blob/84c6308b80acc824493a09dd4713ef2e4333c65a/pkgs/development/libraries/mesa/default.nix#L147

But I really don't want a mass rebuild here so this can go to 21.11 easily (at most a simple backport).

BTW: What's the current behavior of directly using another drv in string interpolation in drv? Does it act like `nativeBuildInputs`, `buildInputs`, or neither? Though `mesa` doesn't have trouble here since they are all target agnostic.
https://github.com/NixOS/nixpkgs/blob/84c6308b80acc824493a09dd4713ef2e4333c65a/pkgs/development/libraries/mesa/default.nix#L102-L104

r-burns commented:
> But I really don't want a mass rebuild here so this can go to 21.11 easily (at most a simple backport).

Ah, good point. Branchoff is drawing near so if buildInputs works okay we can stick with that for now.

> What's the current behavior of directly using another drv in string interpolation in drv? Does it act like nativeBuildInputs, buildInputs, or neither?

Splicing doesn't happen for string interpolation and other literal uses, so unless it's explicitly selected from another package set in the top-level callPackage call, it acts like buildInputs (`pkgs.*`).

r-burns commented:
Oops, forgot to merge this before branchoff. Anyway, the current incarnation looks good to me - and we can backport it direclty without going through staging.

Ericson2314 commented:
Can we implement the missing case in `splice.nix` and then do this more normally? That is safe and good to backport too.

oxalica commented:
> Can we implement the missing case in splice.nix and then do this more normally? That is safe and good to backport too.

The former is okay, but adding a new `depsHostHost` to `mesa` would cause a mass rebuild.

r-burns commented:
Ok, I opened a separate PR to add the missing depsHostHost splicing functionality here: https://github.com/NixOS/nixpkgs/pull/147983

Personally I like the idea of fixing this as-is and then switching to depsHostHost splicing in a followup PR to staging, unless there's a technical reason not to.

Ericson2314 commented:
@r-burns I would use `pkgsHostHost` now with `buildInputs` on master, and then switch to `depsHostHost` without manually mucking about with package sets, on staging. How's that?

r-burns commented:
Ping - by now, enough time has passed that https://github.com/NixOS/nixpkgs/pull/147983 is in master and stable. So I think we can do this the "proper way" and do a trivial backport if still desired.

oxalica commented:
I migrated this PR to use now available `depsHostHost`. It compiles. But I'm not sure how to test it.
The difference of instantiated drvs are just -- as the source diff -- move 3 drvs from `buildInputs` to `depsHostHost`.
As long as it still compile, I guess it's fine?

Ericson2314 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -21195,7 +21195,6 @@ with pkgs;
   libGLU = mesa_glu;
 
   mesa = callPackage ../development/libraries/mesa {
-    llvmPackages = llvmPackages_latest;
```
Can you do get `llvmPackages_latest` from spliced packages or `pkgsHostHost`?

oxalica replied:
The test `doesNotRelyOnTargetPlatform` cannot pass if we keep this override in `all-packages.nix`. But passes if we get `llvmPackages_latest` directly from arguments.

Ericson2314 replied:
The original override (implicitly) takes it from `pkgs`, which is indeed wrong. I am saying get it from somewhere else and then it should be fine.

oxalica replied:
> The original override (implicitly) takes it from `pkgs`, which is indeed wrong. I am saying get it from somewhere else and then it should be fine.

Sorry I don't quite understand. Do you ask to use `llvmPackages = pkgsHostHost.llvmPackages_latest` here? I'm not a fan of leaking details (where `llvmPackages` get used in the drv) into `all-packages.nix`.

SuperSandro2000 commented on pkgs/development/libraries/mesa/default.nix:
```diff
@@ -253,6 +264,15 @@ self = stdenv.mkDerivation {
+      doesNotRelyOnTargetPlatform = stdenv.mkDerivation {
+        name = "mesa-does-not-rely-on-target-platform";
+        # Is there any better way to reference cross-compiled `self`?
+        ok = pkgs.pkgsCross.aarch64-multiplatform.buildPackages.mesa == pkgs.mesa;
```
```suggestion
        ok = pkgsCross.aarch64-multiplatform.buildPackages.mesa == mesa;
```

SuperSandro2000 commented on pkgs/development/libraries/mesa/default.nix:
```diff
@@ -18,6 +18,7 @@
 , enablePatentEncumberedCodecs ? true
 , libclc
 , jdupes
+, pkgs
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/libraries/mesa/default.nix:
```diff
@@ -39,6 +40,10 @@ let
   branch  = versions.major version;
 
+  # We cannot set this in `callPackage` arguments of `all-packages.nix`.
+  # Or LLVM packages will not be splited.
```
```suggestion
  # Or LLVM packages will not be spliced.
```

oxalica committed:
mesa: move LLVM for OpenCL to depsHostHost

Mesa use LLVM for OpenCL, which is both running on and targeting
hostPlatform, thus depsHostHost.

Consider nix-shell with crossSystem for cross platform development,
and have some GUI nativeBuildInputs using mesa. These programs are
expected to run on buildPlatform, and shouldn't depend on crossSystem,
which is buildPackages.mesa's targetPlatform.

oxalica commented:
Rebased. Ready for another review.

<details>

<summary>Diff of <code>nix show-derivation .#mesa</code></summary>

```diff
2c2
<   "/nix/store/k2fr6jiak5knhd0njsmid4ayv10ghjyv-mesa-22.3.2.drv": {
---
>   "/nix/store/c571blqcj2jg9dvrsjdrnm6lbbrgjjam-mesa-22.3.2.drv": {
11c11
<       "buildInputs": "/nix/store/hf4c5sf2w21ab0kwavbn0z9z14zip866-expat-2.5.0-dev /nix/store/ww46d521s4ppyijdlvkmpvxlykfmddda-llvm-14.0.6-dev /nix/store/a6bpzlvh4vrqhjg16n2h9lji6fcbfw4l-libglvnd-1.5.0-dev /nix/store/bsd5gvw94j8glk6f00q4236g8py487vw-xorgproto-2021.5 /nix/store/qjnmrid3d9q85gv6cgj6h0p72af21w79-libX11-1.8.1-dev /nix/store/ab1182v9hbh3wcwxkip3sndgcb4vfmky-libXext-1.3.4-dev /nix/store/dyafqr1ghmkiwsv3yz5ngbzl6n8smmi1-libxcb-1.14-dev /nix/store/zqmakkalal5591mbrj7lf70pfw9fdwl2-libXt-1.2.1-dev /nix/store/4vpbpgrymlhkrdbixbmq8xy0fqihzc5v-libXfixes-6.0.0-dev /nix/store/s7j2zn0rasb3zhhha4fzp3zzyy372nha-libxshmfence-1.3-dev /nix/store/v1i6li5lqkpdjqxv8hvv6kr4zaxmhdx7-libXrandr-1.5.2-dev /nix/store/s7dns7drgskk7xpr4s3kvqcc745dg9js-libffi-3.4.4-dev /nix/store/hvz4f4n3rrcjj4vklbinpjy53cj58z1y-libvdpau-1.5-dev /nix/store/ss604a0wwmsbys09pg09hggpwcgh8ac9-libelf-0.8.13 /nix/store/g4gxrixgx2lja2ippkk8i33z4fysdrlw-libXvMC-1.0.12-dev /nix/store/61nh06456bsahhlp32rv68338iy68pfa-libpthread-stubs-0.4 /nix/store/rrqpfz324x8835wvn1x99d84j8mzv2wd-openssl-3.0.7-dev /nix/store/k64d5w2pzvc8w0pli1m2gr48zn6jj5ml-wayland-1.21.0-dev /nix/store/afngscqcq2jwfkr4pb0r2ipk1sjg4paz-wayland-protocols-1.31 /nix/store/xfb8fvcdz6pbfmbvjphi5qpa66zfs8w7-libomxil-bellagio-0.9.3 /nix/store/fls4dahy09j92zj38bp6bd62hl7c8dgg-libva-minimal-2.16.0-dev /nix/store/y17mqcq9dn829g8864276s633hsvla08-libclc-14.0.6 /nix/store/7vf9kzvc14hy70c8yvd6vbn88358vv3j-clang-wrapper-14.0.6 /nix/store/zpra5ys6xafs8465bbfcf0q3wy9y0bbk-clang-14.0.6-dev /nix/store/5jnh9awh5ywi4sxcxx6i0l8f1zjjkifp-rustc-1.66.0 /nix/store/9c96x1m0q8i8i87w4i8y80xn114nv69q-rust-bindgen-0.61.0 /nix/store/lyx23zfi9w028w0rnzs55gbrpg2c9cli-SPIRV-LLVM-Translator-14.0.0 /nix/store/zybvg883al1zf2mppy301mv0xd468nx3-valgrind-3.20.0-dev",
---
>       "buildInputs": "/nix/store/hf4c5sf2w21ab0kwavbn0z9z14zip866-expat-2.5.0-dev /nix/store/ww46d521s4ppyijdlvkmpvxlykfmddda-llvm-14.0.6-dev /nix/store/a6bpzlvh4vrqhjg16n2h9lji6fcbfw4l-libglvnd-1.5.0-dev /nix/store/bsd5gvw94j8glk6f00q4236g8py487vw-xorgproto-2021.5 /nix/store/qjnmrid3d9q85gv6cgj6h0p72af21w79-libX11-1.8.1-dev /nix/store/ab1182v9hbh3wcwxkip3sndgcb4vfmky-libXext-1.3.4-dev /nix/store/dyafqr1ghmkiwsv3yz5ngbzl6n8smmi1-libxcb-1.14-dev /nix/store/zqmakkalal5591mbrj7lf70pfw9fdwl2-libXt-1.2.1-dev /nix/store/4vpbpgrymlhkrdbixbmq8xy0fqihzc5v-libXfixes-6.0.0-dev /nix/store/s7j2zn0rasb3zhhha4fzp3zzyy372nha-libxshmfence-1.3-dev /nix/store/v1i6li5lqkpdjqxv8hvv6kr4zaxmhdx7-libXrandr-1.5.2-dev /nix/store/s7dns7drgskk7xpr4s3kvqcc745dg9js-libffi-3.4.4-dev /nix/store/hvz4f4n3rrcjj4vklbinpjy53cj58z1y-libvdpau-1.5-dev /nix/store/ss604a0wwmsbys09pg09hggpwcgh8ac9-libelf-0.8.13 /nix/store/g4gxrixgx2lja2ippkk8i33z4fysdrlw-libXvMC-1.0.12-dev /nix/store/61nh06456bsahhlp32rv68338iy68pfa-libpthread-stubs-0.4 /nix/store/rrqpfz324x8835wvn1x99d84j8mzv2wd-openssl-3.0.7-dev /nix/store/k64d5w2pzvc8w0pli1m2gr48zn6jj5ml-wayland-1.21.0-dev /nix/store/afngscqcq2jwfkr4pb0r2ipk1sjg4paz-wayland-protocols-1.31 /nix/store/xfb8fvcdz6pbfmbvjphi5qpa66zfs8w7-libomxil-bellagio-0.9.3 /nix/store/fls4dahy09j92zj38bp6bd62hl7c8dgg-libva-minimal-2.16.0-dev /nix/store/zybvg883al1zf2mppy301mv0xd468nx3-valgrind-3.20.0-dev",
15c15
<       "debug": "/nix/store/p73yynva9w39i9f6v89vw3slzbmayiy6-mesa-22.3.2-debug",
---
>       "debug": "/nix/store/acr7xcvwj3liip7bcwp6z8h2v130mr7g-mesa-22.3.2-debug",
20c20
<       "depsHostHost": "",
---
>       "depsHostHost": "/nix/store/y17mqcq9dn829g8864276s633hsvla08-libclc-14.0.6 /nix/store/7vf9kzvc14hy70c8yvd6vbn88358vv3j-clang-wrapper-14.0.6 /nix/store/zpra5ys6xafs8465bbfcf0q3wy9y0bbk-clang-14.0.6-dev /nix/store/5jnh9awh5ywi4sxcxx6i0l8f1zjjkifp-rustc-1.66.0 /nix/store/9c96x1m0q8i8i87w4i8y80xn114nv69q-rust-bindgen-0.61.0 /nix/store/lyx23zfi9w028w0rnzs55gbrpg2c9cli-SPIRV-LLVM-Translator-14.0.0",
24c24
<       "dev": "/nix/store/rpqq377dr5nr8dr0nxcjwmp81gl7qmzc-mesa-22.3.2-dev",
---
>       "dev": "/nix/store/1j0d4zb4798f0rf9rkm5v48fixid999a-mesa-22.3.2-dev",
27,28c27,28
<       "drivers": "/nix/store/xjhpr3ivqp65q66y58r7a58pyavgs3bd-mesa-22.3.2-drivers",
<       "driversdev": "/nix/store/rp23p1bn862bs59igparvhlpgs9b4s1g-mesa-22.3.2-driversdev",
---
>       "drivers": "/nix/store/n42w1whg8n1zxrrs1l5w5j7miqxdpnid-mesa-22.3.2-drivers",
>       "driversdev": "/nix/store/jr4h5056344qwlk4vx0s2gv0i605b5x1-mesa-22.3.2-driversdev",
32,34c32,34
<       "opencl": "/nix/store/98s9f0yycbvqzm62k46hg6hazq46s4cf-mesa-22.3.2-opencl",
<       "osmesa": "/nix/store/cxq3cj33rhc2srdxdl1h1k7vxclh1563-mesa-22.3.2-osmesa",
<       "out": "/nix/store/pkw4jwgxnzsxds5xdd9jv2r0ji22hikv-mesa-22.3.2",
---
>       "opencl": "/nix/store/rbcg0dvd5h2a64sfppwhg9h9sly4r880-mesa-22.3.2-opencl",
>       "osmesa": "/nix/store/wqpvdnpl65q45xyg3acy76mdkmds8ddh-mesa-22.3.2-osmesa",
>       "out": "/nix/store/k95z2i313lb26i28wi8wvicskylc47wi-mesa-22.3.2",
199c199
<         "path": "/nix/store/p73yynva9w39i9f6v89vw3slzbmayiy6-mesa-22.3.2-debug"
---
>         "path": "/nix/store/acr7xcvwj3liip7bcwp6z8h2v130mr7g-mesa-22.3.2-debug"
202c202
<         "path": "/nix/store/rpqq377dr5nr8dr0nxcjwmp81gl7qmzc-mesa-22.3.2-dev"
---
>         "path": "/nix/store/1j0d4zb4798f0rf9rkm5v48fixid999a-mesa-22.3.2-dev"
205c205
<         "path": "/nix/store/xjhpr3ivqp65q66y58r7a58pyavgs3bd-mesa-22.3.2-drivers"
---
>         "path": "/nix/store/n42w1whg8n1zxrrs1l5w5j7miqxdpnid-mesa-22.3.2-drivers"
208c208
<         "path": "/nix/store/rp23p1bn862bs59igparvhlpgs9b4s1g-mesa-22.3.2-driversdev"
---
>         "path": "/nix/store/jr4h5056344qwlk4vx0s2gv0i605b5x1-mesa-22.3.2-driversdev"
211c211
<         "path": "/nix/store/98s9f0yycbvqzm62k46hg6hazq46s4cf-mesa-22.3.2-opencl"
---
>         "path": "/nix/store/rbcg0dvd5h2a64sfppwhg9h9sly4r880-mesa-22.3.2-opencl"
214c214
<         "path": "/nix/store/cxq3cj33rhc2srdxdl1h1k7vxclh1563-mesa-22.3.2-osmesa"
---
>         "path": "/nix/store/wqpvdnpl65q45xyg3acy76mdkmds8ddh-mesa-22.3.2-osmesa"
217c217
<         "path": "/nix/store/pkw4jwgxnzsxds5xdd9jv2r0ji22hikv-mesa-22.3.2"
---
>         "path": "/nix/store/k95z2i313lb26i28wi8wvicskylc47wi-mesa-22.3.2"
```

</details>
